tket2/src/serialize/pytket/encoder.rs

```diff
@@ -169,21 +188,345 @@ impl JsonEncoder {
+    /// if that was stored in the metadata.
+    outputs: Option<Vec<RegisterUnit>>,
+    /// Map each bit wire to a TKET1 register element.
+    bit_to_reg: HashMap<Wire, RegisterUnit>,
+    /// Registers defined in the metadata, but not present in the circuit
+    /// inputs.
+    unused_registers: VecDeque<RegisterUnit>,
+}
+
+impl BitTracker {
+    /// Create a new [`BitTracker`] from the bit inputs of a [`Circuit`].
+    /// Reads the [`METADATA_B_REGISTERS`] metadata entry with preset pytket bit register names.
+    ///
+    /// If the circuit contains more bit inputs than the provided list,
+    /// new registers are created for the remaining bits.
+    ///
+    /// TODO: Compute output bit permutations when finishing the circuit.
+    pub fn new(circ: &Circuit<impl HugrView>) -> Self {
+        let mut tracker = BitTracker::default();
+
+        if let Some(input_regs) = circ
+            .hugr()
+            .get_metadata(circ.parent(), METADATA_B_REGISTERS)
+        {
+            tracker.inputs = serde_json::from_value(input_regs.clone()).unwrap();
+        }
+        let output_regs = circ
+            .hugr()
+            .get_metadata(circ.parent(), METADATA_B_OUTPUT_REGISTERS)
+            .map(|regs| serde_json::from_value(regs.clone()).unwrap());
+        if let Some(output_regs) = output_regs {
+            tracker.outputs = Some(output_regs);
+        }
+
+        let bit_input_wires = circ.units().filter_map(|u| match u {
+            (CircuitUnit::Wire(w), _, ty) if ty == BOOL_T => Some(w),
+            _ => None,
+        });
+
+        let mut unused_registers: HashSet<RegisterUnit> = tracker.inputs.iter().cloned().collect();
+        for (i, wire) in bit_input_wires.enumerate() {
+            // If the input is not used in the circuit, ignore it.
+            if circ
+                .hugr()
+                .linked_inputs(wire.node(), wire.source())
+                .next()
+                .is_none()
+            {
+                continue;
+            }
+
+            // Use the given input register names if available, or create new ones.
+            if let Some(reg) = tracker.inputs.get(i) {
+                unused_registers.remove(reg);
+                tracker.bit_to_reg.insert(wire, reg.clone());
+            } else {
+                let reg = tracker.add_bit_register(wire).clone();
+                tracker.inputs.push(reg);
+            };
+        }
+
+        // If a register was defined in the metadata but not used in the circuit,
+        // we keep it so it can be assigned to an operation output.
+        tracker.unused_registers = unused_registers.into_iter().collect();
+
+        tracker
+    }
+
+    /// Add a new register unit for a bit wire.
+    pub fn add_bit_register(&mut self, wire: Wire) -> &RegisterUnit {
+        let reg = match self.unused_registers.pop_front() {
+            Some(reg) => reg,
+            None => RegisterUnit("c".to_string(), vec![self.bit_to_reg.len() as i64]),
+        };
```

Comment:
not covered, also might be a problem for targets that enforce bit register max sizes for the problems mentioned in the qubit version